exla/lib/exla/op.ex

```diff
@@ -198,20 +198,45 @@ defmodule EXLA.Op do
     %{op | ref: result_ref}
   end
 
-  def is_nan(op, type) do
-    %{ref: ref} =
-      case type do
-        {t, _} when t in [:f, :bf] ->
-          op
+  def is_nan(op, {t, _}, _shape, _axes, _state) when t in [:f, :bf] do
+    %{ref: ref} = op
+    result_ref = EXLA.NIF.is_nan(ref) |> unwrap!()
+    %{op | ref: result_ref}
+  end
 
-        _ ->
-          convert_element_type(op, Nx.Type.to_floating(type))
-      end
+  def is_nan(_op, _type, shape, axes, %{builder: builder}) do
+    # For non-floating types, we can just return
+    # a boolean 0 tensor in the output shape
+    builder
+    |> constant_r0(0, {:u, 8})
+    |> reshape(Tuple.duplicate(1, tuple_size(shape)))
+    |> broadcast_in_dim(shape, List.to_tuple(axes))
+  end
+
+  def is_infinity(%{ref: ref} = op, {:c, _}, _shape, _axes, _state) do
+    re_part = ref |> EXLA.NIF.real() |> unwrap!() |> EXLA.NIF.is_infinity() |> unwrap!()
+    im_part = ref |> EXLA.NIF.imag() |> unwrap!() |> EXLA.NIF.is_infinity() |> unwrap!()
+
+    result_ref = EXLA.NIF.bitwise_or(re_part, im_part, {}) |> unwrap!()
 
-    result_ref = EXLA.NIF.is_nan(ref) |> unwrap!()
     %{op | ref: result_ref}
   end
 
+  def is_infinity(op, {t, _}, _shape, _axes, _state) when t in [:f, :bf] do
+    %{ref: ref} = op
+    result_ref = EXLA.NIF.is_infinity(ref) |> unwrap!()
+    %{op | ref: result_ref}
+  end
+
+  def is_infinity(_op, _type, shape, axes, %{builder: builder}) do
+    # For non-floating types, we can just return
```

Comment:
@josevalim is this close to what you wanted?
I went down this route because multiplication would also need to be followed by conversion to u8 afterwards

Reply:
Also: should we merge is_nan and is_infinity into a is_non_finite function here that receives the operation as an argument? One is the copy-paste of the other aside from the NIF function itself

Reply:
Looks good! Should we encapsulate it in a function to reuse it?

Reply:
Yeah, I think I'll reuse the same code of is_nan and is_infinity